### PR TITLE
Add support for new feature syntax (RFC 3143)

### PR DESCRIPTION
**Pull request by ehuss**
This adds support for the new feature syntax described in [RFC 3143](https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html).  There are two new feature values that are allowed in the `[features]` table: weak dependencies (`foo?/feat-name`) and namespaced dependencies (`dep:bar`).  When these style of features are detected, they are placed in the index in a separate field called `features2` to prevent older versions of cargo from breaking on them. Additionally, it sets a new version field called `v` to indicate that this new field is present, which prevents versions starting at 1.51 from selecting those versions.

### CODE REVIEW

File: src/git.rs
Comment (jtgeibel) [inline src/git.rs:95]: I'm not sure if there is a precedent for this (having editions apply beyond `rustc`), but I wonder if upon publish we should require the crate to be `edition = "2021"` if `features2` is used.

With this requirement we would ensure that if older versions of cargo do select a crate with features it doesn't understand, then the compilation would bail out early due to the new edition, rather than potentially completing successfully but with an unexpected selection of features.

Comment (ehuss) [reply src/git.rs:95]: Hm, that's an interesting idea.  However, I'm not sure if it will help in enough situations to justify adding the restriction.  If the package required the feature to be set, it would fail during resolution before the package gets downloaded.  If somehow it makes it through resolution, I think at least versions older than 1.27 will fail to parse the manifest.  We are also planning to suggest only using this new syntax in a major version bump which should further limit the impact.  I would kinda prefer to avoid adding more restrictions if possible.  Do you feel strongly about trying to limit the confusion here?

File: src/models/krate.rs
Comment (pksunkara) [inline src/models/krate.rs:509]: Why is this not valid? `dep:foo/bar` should also be valid IMO

Comment (ehuss) [reply src/models/krate.rs:509]: We decided not to allow combining the the two syntaxes to simplify it.  There is more information at https://github.com/rust-lang/cargo/pull/9574 about that change.

**Comment by djc** [comment]
Gentle ping? I think this should be labeled as waiting on review again?

File: src/models/krate.rs
Comment (jhpratt) [inline src/models/krate.rs:308]:
```suggestion
        match name.split_once('/') {
            Some((dep, dep_feat)) => {
                let dep = dep.strip_suffix('?').unwrap_or(dep);
                Crate::valid_feature_prefix(dep) && Crate::valid_feature_name(dep_feat)
            }
            None => Crate::valid_feature_name(name.strip_prefix("dep:").unwrap_or(name)),
        }
```

**Review by jhpratt — COMMENTED** [review]
I can't attest to the validity of the changes as a whole; I'm not familiar with the internals. The parsing of the feature seems correct. Just one minor nit on style.

**Comment by Turbo87** [comment]
FWIW we discussed this on Discord a few days ago: https://discord.com/channels/442252698964721669/448525639469891595/920823792631152691

the tl;dr is that we show features on the UI these days and we need to verify that this still works correctly with this new syntax

**Comment by ehuss** [comment]
> the tl;dr is that we show `features` on the UI these days and we need to verify that this still works correctly with this new syntax

I'm not sure how you'd like to verify that.  I could add something that did a database query in the `features_version_2` test if you'd like.

The feature values themselves don't seem to be displayed in the UI:

image

I can confirm from the API that it is there:

```
curl http://localhost:4200/api/v1/crates/foo | jq .versions[0].features
{
  "f1": [
    "dep:bar"
  ],
  "f2": [
    "bar?/bar-feat"
  ]
}
```

And via psql:

```
cargo_registry=# select * from versions;
 id | crate_id |  num  |         updated_at         |         created_at         | downloads |                   features                   | yanked | license | crate_size | published_by
----+----------+-------+----------------------------+----------------------------+-----------+----------------------------------------------+--------+---------+------------+--------------
  1 |        1 | 0.1.0 | 2021-12-22 23:49:25.411573 | 2021-12-22 23:49:25.411573 |         0 | {}                                           | f      | MIT     |        702 |            1
  2 |        2 | 0.1.0 | 2021-12-22 23:51:35.513816 | 2021-12-22 23:51:35.513816 |         0 | {"bar-feat": []}                             | f      | MIT     |        719 |            1
  4 |        1 | 0.2.0 | 2021-12-22 23:56:47.374767 | 2021-12-22 23:56:47.374767 |         0 | {"f1": ["dep:bar"], "f2": ["bar?/bar-feat"]} | f      | MIT     |        772 |            1
(3 rows)
```

**Review by Turbo87 — APPROVED** [review]
> The feature values themselves don't seem to be displayed in the UI

you're absolutely right. now that I finally had time to take a detailed look at the RFC and PR changes, I agree that the frontend shouldn't be affected by this. since the split between `features` and `features2` only happens for the index, but is otherwise kept in its merged form I don't see any issues with it.

nice work! and sorry for the huge review delay... 😞 

@pietroalbini @jtgeibel unless you have any objections I plan to merge this on 2021-12-30, or earlier if you both signal your approval until then :)

**Comment by Turbo87** [comment]
rebased, merged and deployed. thanks for working on this @ehuss!

**Pull request merged**
